RAT/classlist.py

```diff
@@ -0,0 +1,202 @@
+"""The classlist module. Contains the ClassList class, which defines a list containing instances of a particular class.
+"""
+
+import collections
+import collections.abc
+import contextlib
+import tabulate
+from typing import Any, Dict, Iterable, Sequence, Union
+
+
+class ClassList(collections.UserList):
+    """List of instances of a particular class.
+
+    This class subclasses collections.UserList to construct a list intended to store ONLY instances of a particular
+    class, given on initialisation. Any attempt to introduce an object of a different type will raise a ValueError.
+    The class must be able to accept attribute values using keyword arguments. In addition, if the class has the
+    attribute "name", the ClassList will ensure that all objects within the ClassList have different names. It is then
+    possible to use the name of an object in the .remove(), .count(), and .index() routines in place of the full
+    object. Due to the requirement of unique names, the multiplication operators __mul__, __rmul__, and __imul__ have
+    been disabled, since they cannot allow for unique names by definition.
+
+    We extend the UserList class to enable objects to be added and modified using just the keyword arguments, enable
+    the object name to be used in place of the full object, and ensure all elements are of the specified type, with
+    unique names defined.
+
+    Parameters
+    ----------
+    init_list : Sequence [object] or object
+        An instance, or list of instance(s), of the class to be used in this ClassList.
+    empty_list : bool, optional
+        If true, do not initialise the list with the contents of init_list, use an empty list instead
+        (default is False).
+
+    Raises
+    ------
+    ValueError
+        Raised if the input list is empty.
+    """
+
+    def __init__(self, init_list: Union[Sequence[object], object], empty_list: bool = False) -> None:
```

Comment:
My thinking on this one was that you need to make sure that self._class_handle is defined on initialisation, and cannot be unset. If it is unset, you can't use the append() or insert() routines since we don't know where the arguments should go. Can you think of any other way to make sure the class for the list is defined properly?